lifelines/utils/ShiftedBetaGeometric.py

```diff
@@ -0,0 +1,726 @@
+
+        # update log-likelihood with regularization val.
+        log_like -= l2_reg
+
+        # Get the full alpha and beta for each sample and store the values in
+        # ndarrays of shape (n_samples, )
+        alpha, beta = self._compute_alpha_beta(X, wa, wb)
+
+        # loop over data and add the contribution to the log-likelihood from
+        # each sample in the dataset.
+        # Notice that once we have alpha and beta for each sample we do not
+        # need the feature matrix to compute the contribution to the
+        # log-likelihood.
+        retention_probs = self._log_retention_stats(alpha, beta, age)
+
+        # Contribution to log likelihood based on status (alive vs. dead)
+        log_like += np.sum(retention_probs[0][np.where(alive == 0)[0]])
+        log_like += np.sum(retention_probs[1][np.where(alive == 1)[0]])
+
+        # Negative log_like since we will use scipy's minimize object.
+        return -log_like
+
+    def fit(self, X, age, alive, restarts=1):
+        """
+        Method responsible for the learning step it takes all the relevant data
+        as argument as well as the number of restarts with random seeds to
+        perform. While restarting with other seeds sounds like a good idea the
+        model has proven to be fairly stable and this may be removed in the
+        future.
+
+        :param X: ndarray of shape (n_samples, n_features)
+            The feature matrix
+
+        :param age: ndarray of shape (n_samples, )
+            An array with the age of each individual.
+
+        :param alive: ndarray of shape (n_samples, )
+            An array with
+
+        :param restarts: int
+            Number of times to run the optimization procedure with different seeds
+        """
+
+        # Make sure ages are all non negative...
+        min_age = min(age)
+        if min(age) < 1:
+            raise ValueError("All values of age must be equal or greater to "
+                             "one. The minimum value of "
+                             "{} was found.".format(min_age))
+
+        # Make sure alive is either zero or one!
+        alive_vals = set(alive)
+        if alive_vals != {0, 1}:
+            raise ValueError('Values for alive must be either zero or one. A '
+                             'value of '
+                             '{} was found.'.format(list(alive_vals - {0, 1})))
+
+        # The amount of free space to create when printing based on the total
+        # amount of restarts.
+        print_space = int(max(int(log10(restarts)) + 1, 5))
+
+        # store the number of samples we are dealing with
+        self.n_samples = age.shape[0]
+
+        # Now that we have X we can calculate the number of parameters we need
+        n_params = X.shape[1]
+
+        # --- Optimization Starting Points
+        # Generate random starting points for optimization step.
+        initial_guesses = 0.1 * np.random.randn(restarts, 2 * n_params) - 0.01
```

Comment:
This version extends the original SBG with a log-linear model for predictors. So weights centred at zero are not a natural choice. The choice of `0.01` was purely driven by numerical convenience when I was experimenting with it. I am open to suggestions.